(WIP of #41002) Add support for Reference Handling on Deserialization.

jozkee opened the pull request:
Add support for metadata properties ($id, $ref and $values) used for reference handling on Deserialization.

This is a work in progress and as such, some things in the implementation are still missing, like correctly handling exception messages, property names, etc.

The main purpose of this PR for me is to validate with the area owners if I am on the right path on the prototype/implementation.

This PR only covers the Read or Deserializaiton part, I will address Write/Serialization in a separate PR.

jozkee committed:
Add support for metadata properties ($id, $ref and $values) used for Reference Handling on Deserialization.

steveharter commented on src/System.Text.Json/src/System/Text/Json/Serialization/ReferenceHandlingOnDeserialize.cs:
```diff
@@ -0,0 +1,21 @@
+namespace System.Text.Json
+{
+    /// <summary>
+    /// This enum defines the various ways the <see cref="Utf8JsonReader"/> can deal with comments.
```
Summary is not correct (copied from another area I presume)

steveharter commented on src/System.Text.Json/src/System/Text/Json/Serialization/ReadStack.cs:
```diff
@@ -18,6 +18,74 @@ internal struct ReadStack
+            }
+        }
+
+        public void SetReference(string key, object value)
```
AddReference() may be more descriptive

steveharter commented on src/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.HandleReference.cs:
```diff
@@ -0,0 +1,162 @@
+    }
+
+    // Preserve Reference Handling
+    internal enum MetadataPropertyName
```
Just "MetadataProperty"?

steveharter commented on src/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.HandleReference.cs:
```diff
@@ -0,0 +1,162 @@
+                // since pending objects are initialized when a non-metadata property is found, having zero pending tasks means that $ref is not the first property in the object.
+                if (state.PendingTasksCount == 0)
+                {
+                    throw new JsonException("Properties other than $ref are not allowed in reference objects.");
```
Does Json.NET require $ref as the one and only first property? Also it seems like other metadata that we don't recognize could come first. In general, we should align to Json.NET behavior, and also have tests that actually interop with Json.NET syntax (or use the same input data) -- we do this quite a bit in the reader\writer tests and a little in the serializer custom converter tests.

Jozkee replied:
Json.Net allows other properties to be before $ref but I dont think we shall allow that. However you are right, I still need to validate that other known metadata properties ($id) are not found before $ref.

steveharter commented on src/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.cs:
```diff
@@ -43,15 +43,32 @@ public static partial class JsonSerializer
 
                     JsonTokenType tokenType = reader.TokenType;
 
+                    if (options.ReadReferenceHandling == ReferenceHandlingOnDeserialize.PreserveDuplicates && tokenType == JsonTokenType.PropertyName)
```
This loop is extremely perf-sensitive. Please avoid comparing against the same tokenType more than once.

steveharter commented on src/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.cs:
```diff
@@ -70,16 +87,46 @@ public static partial class JsonSerializer
                         else if (readStack.Current.IsProcessingDictionaryOrIDictionaryConstructible())
                         {
-                            HandleStartDictionary(options, ref readStack);
+                            readStack.EnqueueInitTask(InitTaskType.Dictionary);
```
Please measure perf of this. This feature should not degrade perf when the feature is not active and this method will always create a dictionary.

steveharter commented on src/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.cs:
```diff
@@ -70,16 +87,46 @@ public static partial class JsonSerializer
                     else if (tokenType == JsonTokenType.EndObject)
                     {
-                        if (readStack.Current.Drain)
+                        if (readStack.Current.LastMetaProperty == MetadataPropertyName.Ref)
```
Please add to helper method as to not clutter of the main loop.

steveharter commented on src/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.cs:
```diff
@@ -70,16 +87,46 @@ public static partial class JsonSerializer
+                                }
+                            }
+                        }
+                        else if (readStack.Current.HandleWrappingBrace)
```
I don't understand why we need `HandleWrappingBrace`

Jozkee replied:
When you have to add an $id to an array, the array gets transformed into an object:
```
{
    "$id": "1",
    "$values": []
}
```

So this flag is meant to set to true when this "Preserved Array" is passed in order to avoid handling the close curly brace as the end of an object.

steveharter commented on src/System.Text.Json/src/System/Text/Json/Serialization/ReferenceHandlingOnDeserialize.cs:
```diff
@@ -0,0 +1,21 @@
+    /// <summary>
+    /// This enum defines the various ways the <see cref="Utf8JsonReader"/> can deal with comments.
+    /// </summary>
+    public enum ReferenceHandlingOnDeserialize
```
Doesn't this also apply to whether we serialize the metadata?

Jozkee replied:
Sorry for not to update the original issue, I was discussing with @ahsonkhan offline about splitting `ReferenceHandling` into two separate properties:

`ReferenceHandlingOnDeserialize` and `ReferenceHandlingOnSerialize` and each one would be used exclusively for deserialize or serialize respectively.

steveharter commented on src/System.Text.Json/src/System/Text/Json/Serialization/ReferenceHandlingOnDeserialize.cs:
```diff
@@ -0,0 +1,21 @@
+        /// <summary>
+        /// Indicates to the deserialier to ignore metadata properties in the payload.
+        /// </summary>
+        IgnoreMetadata,
```
I believe earlier discussions is that we would have a default value of "Error" which would throw on deserialize and of course not emit metadata on serialize.

Jozkee replied:
I don't recall said discussion about Error for Deserialize, I do recall that we would throw by default when serializing a refence loop.

steveharter replied:
If IgnoreMetadata is the default, does it throw when there are cycles (like today?)

Jozkee replied:
I think you are confussing Serialize with Deserialize, infinite loops or cycles are not a problem while **Deserializeing/Reading** because you are reading a string which is finite. And as of today, there is no way to have a cycle in a json payload.

The cycles are just a problem when you are **Serializing/Writing** because without a handling you would never finish writing your object graph.

steveharter replied:
Sorry I had a typo -- I meant serialize. Yes the infinite loop only occurs on serialize today. 

I assume this enum also applies to serialize (whether or not metadata is included on serialize) but has "Deserialize" in its name (unless I'm missing some serializer-only setting).

Since this new feature enables a graph, which can have cycles, we need to ensure that valid or malformed\crafted JSON never results in an infinite loop on deserialize.

Jozkee replied:
> I assume this enum also applies to serialize

It does not, there will be a serialize-specific enum and a property on JsonSerializerOptions (probably called `WriteReferenceHandling`) meant for Serialization only. I will update the gist that contains the spec in a moment, sorry for any confussion.

jozkee committed:
Refactored the code:
Removed _pendingInitializations list and anything related, I figure out that you can handle the delayed initialization with only with fields.
Removed duplicated conditions, now the code calls HandleDelayedStart in all code paths that is needed.

Jozkee commented:
`System.Text.Json.Tests.JsonSerializerTests.IncompleteContainers`
`System.Text.Json.Serialization.Tests.ExceptionTests.PathForArrayFails`

These tests were expected to fail, they validate that JsonPath is correct when throwing.
Right now the feature is messing with that, so changes need to be made in order to adjust this with JsonPath.

Jozkee commented:
Closing this in order to stick to new approach from #41885
